sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -69,18 +69,58 @@ type ctxKey string
 const bundleKey ctxKey = "beam:bundle"
 const ptransformKey ctxKey = "beam:ptransform"
 
+// context is a caching context since reads & writes are expensive.
+type beamCtx struct {
+	context.Context
+	bundleID, ptransformID string
+}
+
+// Lift the keys value for faster lookups when not available.
+func (ctx *beamCtx) Value(key interface{}) interface{} {
+	switch key {
+	case bundleKey:
+		if ctx.bundleID == "" {
+			if id := ctx.Value(key); id != nil {
+				ctx.bundleID = id.(string)
+			}
+		}
+		return ctx.bundleID
+	case ptransformKey:
+		if ctx.ptransformID == "" {
+			if id := ctx.Value(key); id != nil {
+				ctx.ptransformID = id.(string)
+			}
+		}
+		return ctx.ptransformID
+	}
+	return ctx.Context.Value(key)
+}
+
 // SetBundleID sets the id of the current Bundle.
 func SetBundleID(ctx context.Context, id string) context.Context {
-	return context.WithValue(ctx, bundleKey, id)
+	var ptransformID string
+	if bctx, ok := ctx.(*beamCtx); ok {
```

Comment:
This doesn't handle the case where the provided context isn't the beamCtx, and accordingly ptransformID wouldn't be set. Is this an error? Should we just silently not have a ptransformID?

Reply:
Or is this the way it is because we don't want strict ordering between SetBundleID and SetPTransformID? If that's the case, that's worth documenting.

Reply:
This is not an error.
BundleID will in practice always be set before any ptransformID, but I don't like making that assumption.
This approach is the "silently not have a ptransformId" variant.

If this isn't a beamCtx that's OK, calling this method returns a beamCtx.

If it's not a beamCtx, the Value method is invoked. If a beamCtx is anywhere on the stack, eventually the beamCtx.Value method will be called, which then delegates down the stack if the ctx doesn't have the value cached already, and then it caches it for later to prevent it from going deeper.
The necessity of the cache is partly mitigated by retaining an earlier configured context in the execution node, but the cache is valuable in avoiding unnecessary type assertion overhead.

Reply:
The semantics here is subtle and deserves a comment. I would suggest making a note that checking for beamCtx here is just an optimization. Always using "" as ptransformID would work as well.

Ditto for the other set helper.